src/base_alloc/base_alloc_linear.c

```diff
@@ -160,6 +172,65 @@ void *umf_ba_linear_alloc(umf_ba_linear_pool_t *pool, size_t size) {
     return ptr;
 }
 
+// umf_ba_linear_free() really frees memory only if all allocations from an inactive pool were freed
+void umf_ba_linear_free(umf_ba_linear_pool_t *pool, void *ptr) {
+    util_mutex_lock(&pool->metadata.lock);
+#ifndef NDEBUG
+    ba_debug_checks(pool);
+#endif /* NDEBUG */
+    // check if ptr belongs to pool
+    char *cptr = (char *)ptr;
+    if (cptr >= pool->data &&
+        cptr < ((char *)(pool)) + pool->metadata.pool_size) {
```

Comment:
nit: could you extract this check to a function like 'contains_ptr'?

Reply:
Done